images/virtualization-artifact/pkg/controller/service/vm_operation.go

```diff
@@ -0,0 +1,210 @@
+}
+
+func (s VMOperationService) IsAllowedForVM(vmop *virtv2.VirtualMachineOperation, vm *virtv2.VirtualMachine) bool {
+	if vm == nil {
+		return false
+	}
+	return s.IsApplicableForRunPolicy(vmop, vm.Spec.RunPolicy) && s.IsApplicableForVMPhase(vmop, vm.Status.Phase)
+}
+
+func (s VMOperationService) IsApplicableForRunPolicy(vmop *virtv2.VirtualMachineOperation, runPolicy virtv2.RunPolicy) bool {
+	switch runPolicy {
+	case virtv2.AlwaysOnPolicy:
+		return vmop.Spec.Type == virtv2.VMOPTypeRestart
+	case virtv2.AlwaysOffPolicy:
+		return false
+	case virtv2.ManualPolicy, virtv2.AlwaysOnUnlessStoppedManually:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s VMOperationService) IsApplicableForVMPhase(vmop *virtv2.VirtualMachineOperation, phase virtv2.MachinePhase) bool {
+	if phase == virtv2.MachineTerminating ||
+		phase == virtv2.MachinePending ||
+		phase == virtv2.MachineMigrating {
+		return false
+	}
+	switch vmop.Spec.Type {
+	case virtv2.VMOPTypeStart:
+		return phase == virtv2.MachineStopped || phase == virtv2.MachineStopping
+	case virtv2.VMOPTypeStop, virtv2.VMOPTypeRestart:
+		return phase == virtv2.MachineRunning ||
+			phase == virtv2.MachineDegraded ||
+			phase == virtv2.MachineStarting ||
+			phase == virtv2.MachinePause
+	default:
+		return false
+	}
+}
+
+func (s VMOperationService) IsCompleted(opType virtv2.VMOPType, phase virtv2.MachinePhase) bool {
+	switch opType {
+	case virtv2.VMOPTypeRestart, virtv2.VMOPTypeStart:
+		return phase == virtv2.MachineRunning
+	case virtv2.VMOPTypeStop:
+		return phase == virtv2.MachineStopped
+	default:
+		return false
+	}
+}
+
+// OtherVMOPIsInProgress check if there is at least one VMOP for the same VM in progress phase.
+func (s VMOperationService) OtherVMOPIsInProgress(ctx context.Context, vmop *virtv2.VirtualMachineOperation) (bool, error) {
+	var vmopList virtv2.VirtualMachineOperationList
+	err := s.client.List(ctx, &vmopList, client.InNamespace(vmop.GetNamespace()))
+	if err != nil {
+		return false, err
+	}
+
+	for _, other := range vmopList.Items {
+		// Ignore ourself.
+		if other.GetName() == vmop.GetName() {
+			continue
+		}
+		// Ignore VMOPs for different VMs.
+		if other.Spec.VirtualMachine != vmop.Spec.VirtualMachine {
+			continue
+		}
+		// Return true if other VMOP is in progress.
+		if vmop.Status.Phase == virtv2.VMOPPhaseInProgress {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (s VMOperationService) InProgressReasonForType(vmop *virtv2.VirtualMachineOperation) string {
+	if vmop == nil || vmop.Spec.Type == "" {
+		return ""
+	}
+	switch vmop.Spec.Type {
+	case virtv2.VMOPTypeStart:
+		return vmopcondition.StartInProgress
+	case virtv2.VMOPTypeStop:
+		return vmopcondition.StopInProgress
+	case virtv2.VMOPTypeRestart:
+		return vmopcondition.RestartInProgress
+	}
+	return ""
+}
+
+func (s VMOperationService) IsComplete(vmop *virtv2.VirtualMachineOperation, vm *virtv2.VirtualMachine) bool {
```

Comment:
We already have method IsCompleted. 
Consider deleting one of them. Or reuse

Reply:
Forget delete IsCompleted, done.